File: 0020-valid-parentheses/0020-valid-parentheses.rs
use std::collections::HashMap;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let dictionary = HashMap::from([
            ('(', ')'),
            ('{', '}'),
            ('[', ']'),
        ]);

        let mut stack = Vec::new();
        for ch in s.chars() {
            if dictionary.contains_key(&ch) {
                stack.push(ch);
            } else {
                if stack.is_empty() {
                    return false;
                }
                let last = stack.pop().unwrap();
                if dictionary.get(&last) != Some(&ch) {
                    return false;
                }
            }
        }
        stack.is_empty()
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for bracket in s:
            if bracket == '(' or bracket == '{' or bracket == '[':
                stack.append(bracket)
            else:
                if len(stack) == 0:
                    return False
                lastBracket = stack.pop()
                if lastBracket == '(' and bracket != ')':
                    return False
                elif lastBracket == '{' and bracket != '}':
                    return False
                elif lastBracket == '[' and bracket != ']':
                    return False
        return len(stack) == 0
